src/lib.rs

```diff
@@ -968,8 +1045,66 @@ impl GooseAttack {
                 task_set.weight,
                 weight
             );
-            let mut weighted_sets = vec![index; weight];
-            weighted_task_sets.append(&mut weighted_sets);
+            let weighted_sets = vec![index; weight];
+            total_task_sets += weight;
+            available_task_sets.push(weighted_sets);
+        }
+
+        info!(
+            "allocating GooseTasks to GooseUsers with {:?} scheduler",
+            self.scheduler
+        );
+
+        // Now build the weighted list with the appropriate scheduler.
+        let mut weighted_task_sets = Vec::new();
+        match self.scheduler {
+            GooseScheduler::RoundRobin => {
+                // Allocate task sets round robin.
+                let task_sets_len = available_task_sets.len();
+                loop {
+                    for (i, task_sets) in available_task_sets
+                        .iter_mut()
+                        .enumerate()
+                        .take(task_sets_len)
+                    {
+                        if let Some(task_set) = task_sets.pop() {
+                            debug!("allocating 1 user from TaskSet {}", i);
+                            weighted_task_sets.push(task_set);
+                        }
+                    }
+                    if weighted_task_sets.len() >= total_task_sets {
+                        break;
+                    }
+                }
+            }
+            GooseScheduler::Serial => {
+                // Allocate task sets serially in the weighted order defined.
+                for (task_set, task_sets) in available_task_sets.iter().enumerate() {
```

Comment:
Should use `i` here for index like above or even better use: `task_set_index, task_sets`

Reply:
Yes, I use `task_set_index` elsewhere in the code, so using that consistently now here. Fixed.